text/0000-async.md

```diff
@@ -0,0 +1,774 @@
+    /// the current task for wakeup if the value is not yet available.
+    ///
+    /// # Return value
+    ///
+    /// This function returns:
+    ///
+    /// - `Poll::Pending` if the value is not ready yet.
+    /// - `Poll::Ready(val)` with the result `val` upon completion.
+    ///
+    /// Once an async value has completed, clients should not `poll` it again.
+    ///
+    /// When an async value is not ready yet, `poll` returns `Poll::Pending`.
+    /// The computation will *also* register the interest of the current task in the
+    /// value being produced. For example, if the async value represents the availability
+    /// of data on a socket, then the task is recorded so that when data arrives,
+    /// it is woken up (via `cx.waker()`). Once a task has been woken up,
+    /// it should attempt to `poll` the computation again, which may or may not
+    /// produce a final value at that time.
+    ///
+    /// Note that if `Pending` is returned it only means that the *current* task
+    /// (represented by the argument `cx`) will receive a notification. Tasks
+    /// from previous calls to `poll` will *not* receive notifications.
+    ///
+    /// # Runtime characteristics
+    ///
+    /// `Async` values are *inert*; they must be *actively* `poll`ed to make
+    /// progress, meaning that each time the current task is woken up, it should
+    /// actively re-`poll` pending computations that it still has an interest in.
+    /// Usually this is handled automatically by `async`/`await` notation or
+    /// via adapter methods. Executors ensure that each task is `poll`ed every
+    /// time a future internal to that task is ready to make progress.
+    ///
+    /// The `poll` function is not called repeatedly in a tight loop, but only
+    /// whenever the computation itself is ready to make progress, as signaled via
+    /// `cx.waker()`. If you're familiar with the `poll(2)` or `select(2)`
+    /// syscalls on Unix it's worth noting that async values typically do *not*
+    /// suffer the same problems of "all wakeups must poll all events"; they
+    /// are more like `epoll(4)`.
+    ///
+    /// An implementation of `poll` should strive to return quickly, and should
+    /// *never* block. Returning quickly prevents unnecessarily clogging up
+    /// threads or event loops. If it is known ahead of time that a call to
+    /// `poll` may end up taking awhile, the work should be offloaded to a
+    /// thread pool (or something similar) to ensure that `poll` can return
+    /// quickly.
+    ///
+    /// # Panics
+    ///
+    /// Once an async value has completed (returned `Ready` from `poll`), subsequent
+    /// calls to `poll` may panic, block forever, or otherwise cause bad behavior.
+    /// The `Async` trait itself provides no guarantees about the behavior of
+    /// `poll` after a future has completed.
+    fn poll(self: Pin<Self>, cx: &mut task::Context) -> Poll<Self::Output>;
```

Comment:
One possible concern here is that we may want to [rename `Pin` to `PinMut`](https://github.com/rust-lang/rust/issues/49150#issuecomment-381566909) to make room for a shared pinned reference. Should we try to get that renamed before even landing this as an unstable API to minimize churn?

Reply:
This rename has happened in https://github.com/rust-lang/rust/pull/50497, so the RFC should probably be updated.